---
title: "Osteosarcoma scAtlas"
author: "Yogesh Budhathoki and Matt Cannon"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
format:
    html:
        toc: true
        toc_float: true
        toc_depth: 5
        number_sections: false
        code-fold: true
        code-link: true
        df-print: kable
        embed-resources: true
execute:
    cache: true
knitr:
    opts_chunk:
        cache.lazy: false
        cache.vars: ""
        warning: false
        message: false
lightbox: true
---

```{r setup_main}
#| include: FALSE
library(monocle3)
library(rrrSingleCellUtils)
library(rrrSnvs)
library(Seurat)
library(ggrepel)
library(tidyverse)
library(harmony)
library(cowplot)
library(clustree)
library(data.table)
library(hdf5r)
library(Rmagic)
library(knitr)
library(scATOMIC)
library(copykat)
library(SCENIC)
library(stashPlot)
library(ggalluvial)
library(CellChat)
library(RColorBrewer)
library(gridExtra)
library(DoubletFinder)

knitr::purl("child_qmds/helper_functions.qmd", "helpers.R")
source("helpers.R")


# # list of files in order to keep track of the files order
# read_raw_data.qmd
# helper_functions.qmd
# variants.qmd
# annotate_celltypes.qmd
# annot_cancer_scevan_scatomic.qmd
# annot_cancer_celltype.qmd
# split_tumor_stromal.qmd
# degs_gseaplots.qmd
# regulon.qmd
# helper_functions_correlation.qmd
# groupwise_expr_correlations.qmd
# patient_reanalysis.qmd
# stroma_reclustering.qmd
# main_figures.qmd
# TAMs_analysis.qmd

```

# Load in functions

{{< include child_qmds/helper_functions.qmd >}}

# Set up directory structure
```{bash mkdirs}
for directoryName in \
    output \
    output/figures \
    output/figures/gsea \
    output/seurat_objects/final_combined_sobjs \
    output/figures/tumor_vs_stroma/clustree \
    output/figures/stromal_species_correlations \
    output/figures/tumor_subtypes_correlations \
    output/figures/spatial/spacexr \
    output/figures/spatial/lees \
    output/counts \
    output/metadata \
    output/seurat_objects/sobj_preprocess \
    output/seurat_objects/comb_sobjs \
    output/seurat_objects/harmony_sobjs \
    output/seurat_objects/harmony_sobjs_annot \
    output/seurat_objects/sobj_preprocess_ \
    output/seurat_objects/tumor_vs_stroma \
    output/id_tumor/scATOMIC \
    output/id_tumor/SCEVAN \
    output/id_tumor/snv_calling \
    output/id_tumor/snvs/plots \
    output/id_tumor/celltype \
    output/stromal_species_correlations \
    output/tumor_subtypes_correlations \
    output/spacexr/granular_references/ \
    output/spacexr/granular_references/lee_perms
do
    if [ ! -d ${directoryName} ]
    then
        mkdir -p ${directoryName}
    fi
done

```

# Read in raw data and process to Seurat objects
The metadata for all samples is stored in misc/all_samples.txt.

This reads in the raw data from the hdf5 files and processes them to Seurat objects.

The individual objects are saved to `output/seurat_objects/sobj_preprocess/`

Merged data for each group (as defined in the metadata) are saved to `output/seurat_objects/comb_sobjs/`

Harmonized datasets are saved to `output/seurat_objects/sobj_preprocess_merged/`

{{< include child_qmds/read_raw_data.qmd >}}


# Use SingleR and reference datasets to annotate cell types
The output of this is written to `output/seurat_objects/harmony_sobjs_annot/`

 {{< include child_qmds/annotate_celltypes.qmd >}}

<!-- # Use SNV calling to try to ID tumor cells
Run Matts SNV calling pipeline to try to ID tumor cells

{{< include child_qmds/variants.qmd >}} -->

# Use SCEVAN and scATOMIC to try to identify tumor cells
This will help in splitting the cancer cells from the stroma cells.

{{< include child_qmds/annot_cancer_scevan_scatomic.qmd >}}

# Use SingleR cell annotations to try to ID cancer cells

{{< include child_qmds/annot_cancer_celltype.qmd >}}

# Split cancer vs the stroma

{{< include child_qmds/split_tumor_stromal.qmd >}}

# DEGS and GSEA
Run the differential expression and GSEA

{{< include child_qmds/degs_gseaplots.qmd >}}

<!-- # Investigate the expression correlation between subgroups of the data

{{< include child_qmds/helper_functions_correlation.qmd >}}

{{< include child_qmds/groupwise_expr_correlations.qmd >}}

# Regulon Analysis
Perform the regulon analysis using the SCENIC package. The regulon analysis is

{{< include child_qmds/regulon.qmd >}}

# Stromal Reclustering
Recluster the stromal cells

{{< include child_qmds/stromal_reclustering.qmd >}}

# Patient Reanalysis
Reanalyze the patient samples for validation such as harmony integration validation

{{< include child_qmds/patient_reanalysis.qmd >}}

# render to make figures html

{{< include child_qmds/render_figs.qmd >}} --> -->



# Session info
```{r}
sessionInfo()
```
