```{r library}
library(tidyverse)
library(rrrSingleCellUtils)
library(rrrSnvs)
library(Seurat)
```

```{r load_sample_meta}
samples_to_run <-
    read_csv("allsample_details.csv",
             show_col_types = FALSE) %>%
    filter(species %in% c("patient", "mousemodel") &
           location != "cc_alltumor" &
           (tumor_type == "primary" |
            (tumor_type == "metastatic" & location == "lung")) &
           data_source != "NCI-POB") %>% # We don't have the bam files for these samples
    filter(!sample_name %in% c("SJOS013768_D1", # drop samples with low reads
                               "SJOS016016_D1",
                               "SJOS030605_D1",
                               "SJOS030411_D1",
                               "SJOS030605_D2"))
```


Anand's data is here: /home/gdrobertslab/lab/ExternalData/Patel_lab/
    all samples with SJOS.+
our data and GEO data are here: /home/gdrobertslab/lab/Counts_2/
    All samples with S0.+ and X0.+

Seurat objects in:
/home/gdrobertslab/lab/Analysis/Yogesh/CellTypeAnnRefs/output/seurat_objects/sobj_preprocess/

```{r}
file_loc_key <-
    list("X0" = "/home/gdrobertslab/lab/Counts_2/",
         "S0" = "/home/gdrobertslab/lab/Counts_2/",
         "SC" = "/home/gdrobertslab/lab/Counts_2/",
         "SJ" = "/home/gdrobertslab/lab/ExternalData/Patel_lab/")

ploidy_key <-
    list("human" = "GRCh38",
         "mouse" = "mm10")

control_celltypes <-
    c("Monocytes",
      "Macrophage",
      "Macrophages")

genome_key <-
    list("human" = "/home/gdrobertslab/lab/GenRef/10x-hg38/fasta/genome.fa",
         "mouse" = "/home/gdrobertslab/lab/GenRef/10x-mm10/fasta/genome.fa")

# parallel::mclapply(seq_len(nrow(samples_to_run)),
#                    mc.cores = 10,
#                    function(i) {

#for (i in seq_len(nrow(samples_to_run))) {
for (i in 49:nrow(samples_to_run)) {
    sample_id <- samples_to_run$sample_name[i]
    species <- samples_to_run$organism[i]
    file_loc <- file_loc_key[[substr(sample_id, 1, 2)]]
    ploidy <- ploidy_key[[species]]
    data_source <- samples_to_run$data_source[i]
    method <- samples_to_run$method[i]

    message(sample_id)

    bam_file <- "possorted_genome_bam.bam"
    if (method == "single_nucleus" && data_source == "NCH") {
        bam_file <- "gex_possorted_bam.bam"
    }
    message("loading data")
    s_obj <-
        tenx_load_qc(h5_file = paste0(file_loc,
                                      sample_id,
                                      "/filtered_feature_bc_matrix.h5"),
                     violin_plot = FALSE) %>%
        subset(nCount_RNA <= samples_to_run$ncount_max[i] &
               percent.mt <= samples_to_run$mt_percent[i])
    if (length(Cells(s_obj)) > 100) {
        s_obj <-
            s_obj %>%
            process_seurat(run_umap_dims = 1:50) %>%
            annotate(species = species)
    } else {
        message("Only ",
                length(Cells(s_obj)),
                " cells for ",
                sample_id)
        next
        #return(FALSE)
    }

    message("optimize silhouette")
    opt_res <-
        optimize_silhouette(s_obj,
                            summary_plot = FALSE) %>%
        arrange(desc(sil_vals)) %>%
        pull(res_vals) %>%
        head(1)

    s_obj <- FindClusters(s_obj, resolution = opt_res)

    full_bam_file <-
        paste0(file_loc,
               sample_id,
               "/",
               bam_file)

    if (!file.exists(full_bam_file)) {
        message("No bam file for ",
                sample_id,
                " at ",
                full_bam_file)
        next
        #return(FALSE)

    }

    s_obj$bam_file <- full_bam_file

    s_obj$cell_barcode <- colnames(s_obj)
    s_obj$cell_group <- paste0("cluster_", s_obj$seurat_clusters)
    message("plotting")
    dim_plot_name <-
        DimPlot(s_obj,
                group.by = c("seurat_clusters", "annotations"),
                label = TRUE,
                repel = TRUE,
                ncol = 1)

    ggsave(paste0("output/snvs/", sample_id, "_dim_plot.png"),
           plot = dim_plot_name,
           width = 20,
           height = 15)

    c_b_t <-
        s_obj@meta.data %>%
        select(cell_group, bam_file, cell_barcode)

    if (!dir.exists(paste0("output/snvs/", sample_id))) {
        dir.create(paste0("output/snvs/", sample_id), recursive = TRUE)
    }
    message("getting snp tree")
    snp_tree <-
        tryCatch({
            get_snp_tree(cellid_bam_table = c_b_t,
                         ploidy = ploidy,
                         ref_fasta = genome_key[[species]],
                         min_depth = c(5, 10, 20),
                         temp_dir = paste0("output/snvs/", sample_id),
                         sbatch_base = paste0("output/snvs/", sample_id, "/shell"),
                         slurm_base = paste0("output/snvs/", sample_id, "/slurm"),
                         min_sites_covered = 1000,
                         cleanup = TRUE)
        }, error = function(e) {
            message("Error in snp_tree for ",
                    sample_id,
                    ": ",
                    e)
            #return(NULL)
        })

    # if all the min_depths failed due to low coverage, then the value of snp_tree
    # an empty list of length 0. We change this to NULL for downstream processing
    if (length(snp_tree) == 0) {
        snp_tree <- NULL
    }

    # if (is.null(snp_tree)) {
    #     return(FALSE)
    # }
    qs::qsave(snp_tree,
              paste0("output/snvs/",
                     sample_id, "/",
                     sample_id, "_snp_tree.qs"))
    message("plotting trees")
    for (depth in c("min_depth_5", "min_depth_10", "min_depth_20")) {
        if (!is.null(snp_tree[[depth]])) {
            snp_tree[[depth]]$labels <-
                str_remove(snp_tree[[depth]]$labels, "split_")
            png(paste0("output/snvs/", sample_id, "/", depth, "_", sample_id, ".png"),
                width = 2500,
                height = 2500,
                res = 300)
            plot(snp_tree[[depth]])
            dev.off()
        }
    }
    message("labeling tree groups")
    # Figure out which clusters are more than 50% control celltypes
    normal_clusters <-
        match_celltype_clusters(sobject = s_obj,
                                normal_celltypes = control_celltypes,
                                cluster_col = "seurat_clusters",
                                celltype_col = "annotations") %>%
        paste0("cluster_", .)

    cut_n_groups <- 2

    s_obj$cell_barcode <- NULL
    for (depth in c("min_depth_5", "min_depth_10", "min_depth_20")) {
        if (!is.null(snp_tree[[depth]])) {
            s_obj <-
                label_tree_groups(sobject = s_obj,
                                  dist_tree = snp_tree[[depth]],
                                  group_col_name = "cell_group",
                                  normal_groups = normal_clusters,
                                  cut_n_groups = cut_n_groups,
                                  tumor_call_column = paste0("snp_", depth))
        }
    }
    message("dimplots of calls")
    if (!is.null(snp_tree)) {
        plot_name <-
            DimPlot(s_obj,
                    group.by = c("snp_min_depth_5",
                                "snp_min_depth_10",
                                "snp_min_depth_20"))
        ggsave(paste0("output/snvs/", sample_id, "_dim_plot_calls.png"),
            plot = plot_name,
            width = 20,
            height = 15)
    }

    qs::qsave(s_obj, paste0("output/snvs/", sample_id, ".qs"))

    s_obj@meta.data %>%
        rownames_to_column("cell_barcode") %>%
        select(cell_barcode, starts_with("snp_min_depth")) %>%
        write_tsv(paste0("output/snvs/", sample_id, "_snp_calls.tsv"))
}
    #return(TRUE)
#})
```

```{r}

f420 <-
    qs::qread("/home/gdrobertslab/mvc002/analyses/roberts/24_Ley_barcodes/output/rdata/sobj_list.qs")[[2]]
f420$bam_file <-
    paste0("/home/gdrobertslab/lab/Counts_2/",
           f420$Sample_ID,
           "/possorted_genome_bam.bam")
f420$cell_barcode <- colnames(f420)
f420$cell_group <- f420$seurat_clusters


# imm_cells <- celldex::ImmGenData()
# mouse_rna <- celldex::MouseRNAseqData()

# cell_assign <-
#     SingleR::SingleR(as.SingleCellExperiment(test_sobj),
#                      ref = list(imm_cells,
#                                 mouse_rna),
#                      labels = list(imm_cells$label.main,
#                                    mouse_rna$label.main))

# test_sobj$cell_type <- cell_assign$labels
# test_sobj$cell_score <- cell_assign$scores %>%
#     apply(MARGIN = 1, function(x) max(x, na.rm = TRUE))

c_b_t <-
    f420@meta.data %>%
    select(cell_group, bam_file, cell_barcode)

snp_tree <-
    get_snp_tree(cellid_bam_table = c_b_t,
                 ploidy = "mm10",
                 ref_fasta = "/home/gdrobertslab/lab/GenRef/10x-mm10/fasta/genome.fa",
                 min_depth = c(5, 10, 20),
                 temp_dir = "/gpfs0/scratch/mvc002/test_Y",
                 sbatch_base = "test_Y",
                 slurm_base = "/gpfs0/scratch/mvc002/test_Y/two_sobj",
                 min_sites_covered = 1000,
                 cleanup = TRUE)

qs::qsave(snp_tree, "test_snp_tree.RData")

png("testTwoSobjOutput1.png",
    width = 2500,
    height = 2500,
    res = 300)
plot(snp_tree$min_depth_5)
dev.off()

# Do cell type annotation so I can use macrophages and monocytes as known normal cells
# The idea is that Osteosarcoma cells are unlikely to be mistakenly annotated as macrophages or monocytes

control_celltypes <- c("Monocytes", "Macrophages")

# Figure out which clusters are more than 50% control celltypes
normal_clusters <-
    match_celltype_clusters(sobject = test_sobj,
                            normal_celltypes = control_celltypes,
                            cluster_col = "seurat_clusters",
                            celltype_col = "cell_type")

cut_n_groups <- 2

new_two_sobj <-
    label_tree_groups(sobject = two_sobj,
                      dist_tree = snp_tree,
                      group_col_name = "used_clusters",
                      normal_groups = normal_clusters,
                      cut_n_groups = cut_n_groups)


png("testTwoSobjOutput2.png",
    width = 7000,
    height = 2500,
    res = 300)
DimPlot(new_two_sobj,
        group.by = c("snp_tumor_call",
                     "cell_type"),
        label = TRUE,
        repel = TRUE) +
    NoLegend()
dev.off()

```