---
title: "Human Osteosarcoma Primary - Annotate Tumor Cells"
subtitle: "Cell type annotation reference assembly"
output: html_document
---

This script takes the output files from Primary-CreateData script (comb.qs)
and the Primary-Cluster script (selection.rds) to evaluate intratumoral
heterogeneity and to characterize tumor cell subclusters.

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  dev = "svg",
  fig.width = 4,
  fig.height = 4,
  fig.asp = NULL
)

library(rrrSingleCellUtils)
library(Seurat)
library(ggplot2)
library(ggrepel)
library(tidyverse)
library(stringr)
library(harmony)
library(patchwork)
library(ggpubr)

set.seed(888)

sysfonts::font_add("Arial Narrow", "ARIALN.TTF", bold = "ARIALNB.TTF")
showtext::showtext_auto()

theme_set(theme_pubr())

plot_cols <- c("#D43F3AFF", "#EEA236FF", "#357EBDFF", "#5CB85CFF", "#B8B8B8FF",
  "#9632B8FF",  "#46B8DAFF", "#90302DFF", "#A66D04FF", "#2D577FFF", "#3E7E3EFF",
  "#7D7D7DFF", "#6D1D87FF", "#097F9AFF", "#FF6E6AFF", "#FFBB70FF", "#68A4E3FF",
  "#79D379FF", "#CDCDCDFF", "#BF6BE2FF", "#69D1F3FF")

rDimPlot <- function(object, title = "Unlabeled Plot", subtitle = NA,
  label = T, pt.size = 1, ...) {
  if(length(levels(Idents(object))) < 22) {
    p <- Seurat::DimPlot(
      object = object, 
      label = label, 
      pt.size = pt.size,
      cols = alpha(plot_cols, 0.6),
      ...) +
      labs(
        title = title,
        subtitle = subtitle) +
      theme(legend.position = "none") +
      coord_fixed()
    print(p)
    return(p)
  } else {
    print("Too many identity classes to use this function. Requires <22.")
  }
}

rFeaturePlot <- function(object, features, title = "Unlabeled Plot", subtitle = NA,
  pt.size = 1, order = T, ...) {
  p <- Seurat::FeaturePlot(
    object = object,
    features = features,
    pt.size = pt.size,
    order = order,
    cols = (c("lightgoldenrod", "darkred")),
    ...) +
    labs(
      title = title,
      subtitle = subtitle) +
    coord_fixed()
  print(p)
  return(p)
}
```

# Isolate and subcluster the tumor cells

```{r, fig.width = 8, fig.width = 8}
# Isolate the tumor cells away from the comb Seurat object
selection <- readRDS("HuOsteo/selection.rds")
tumor <- subset(qs::qread("HuOsteo/comb.qs"), cells = selection)

# Normalize and scale tumor dataset
tumor <- tumor %>%
  NormalizeData() %>%
  FindVariableFeatures() %>%
  ScaleData() %>%
  RunPCA(verbose = F)

# Run batch-correction/integration using harmony
tumor <- RunHarmony(tumor, group.by.vars = "src")
tumor <- tumor %>%
  RunUMAP(reduction = "harmony", dims = 1:20) %>%
  FindNeighbors(reduction = "harmony", dims = 1:20)

plot <- list()
opt <- tribble(
  ~res, ~score, ~clusters
)

for(i in 1:9) {
  opt[i, "res"] <- 0.1 * i
  tumor <- FindClusters(tumor, resolution = opt$res[i])
  plot[[i]] <- rDimPlot(tumor, paste("Resolution", opt$res[i]))
  opt[i, "score"] <- silhouette_mean(silhouette_seurat(tumor))
  opt[i, "clusters"] <- length(levels(tumor$seurat_clusters))
}

wrap_plots(plot, ncol = 3)

# Manually enforcing a resolution of 0.2 based on review of clusters
# and potential DEGs. Silhouette performs poorly for scoring intratumor
# heterogeneity.
tumor <- FindClusters(tumor, resolution = 0.2)
clusters <- levels(factor(tumor$seurat_clusters))
names(clusters) <- paste0("Cluster", clusters)
```

# Generate DEG lists for each cluster using a pseudobulk procedure

First, generate DEG tables from pseudobulk clusters, treating all cells
from each cluster from each sample as independent groups of cells/data.
Then, graph each of those DEG datasets using a volcano plot.

```{r}
# Create pseudobulk tables
pseudobulk <- data.frame(gene = rownames(tumor@assays$RNA@counts))
rownames(pseudobulk) <- pseudobulk$gene

for(s in levels(as_factor(tumor$src))) {
  subtumor <- subset(tumor, subset = src == s)
  for(c in clusters) {
    if(length(which(subtumor$seurat_clusters == c)) > 9) {
      subclust <- subset (subtumor, seurat_clusters == c)
      lab <- paste0(s, "-C", c)
      pseudobulk[lab] <- rowSums(subclust@assays$RNA@counts)
    } else {
      print(paste("NOTE: Cluster", c, "from sample", s,
        "has less than 10 cells. This cluster will be skipped."))
    }
  }
}
pseudobulk <- pseudobulk[, -1]

# Create a separate column data table as required for generating DESeq2 input
coldata <- data.frame(sample = colnames(pseudobulk))
rownames(coldata) <- coldata$sample
coldata <- separate(coldata, "sample", c("sample", "cluster"), sep = "-C")
coldata$type <- "single-read"
coldata <- coldata[, -1]

# Perform differential expression analysis between clusters
ds2 <- list()
plots <- list()
for(c in names(clusters)) {
  cluster_number <- clusters[c]
  coldata_targeted <- coldata
  coldata_targeted["cluster"][coldata_targeted["cluster"] == cluster_number] <- "target"
  coldata_targeted["cluster"][coldata_targeted["cluster"] != "target"] <- "other"

  ds2[[c]]$data_sets <- DESeq2::DESeqDataSetFromMatrix(
    countData = pseudobulk,
    colData = coldata_targeted,
    design = ~ cluster
  )

  ds2[[c]]$tests <- DESeq2::DESeq(ds2[[c]]$data_sets)
  ds2[[c]]$results <-
    as.data.frame(DESeq2::results(ds2[[c]]$tests))

  ds2[[c]]$results <-
    ds2[[c]]$results[complete.cases(ds2[[c]]$results), ]
  ds2[[c]]$results <-
    ds2[[c]]$results[order(-abs(ds2[[c]]$results$log2FoldChange)), ]
  labs <- head(ds2[[c]]$results, n = 15)

  plots[[c]][["pb_ds2"]]$volcano <-
    ggplot(ds2[[c]]$results, aes(log2FoldChange, -log10(padj))) +
      geom_point() +
      geom_label_repel(
        data = labs,
        aes(label = rownames(lab)),
        max.overlaps = 20) +
      labs(
        title = paste(c, "Genes"),
        subtitle = "pb_ds2 method")
}
```

# Generate DEG lists and volcano plots using Seurat-Wilcox and ROC DE analysis
```{r}
Idents(tumor) <- tumor$seurat_clusters
deg_test <- c("wilcox", "roc")
scresults <- list()
for(t in deg_test) {
  for(c in names(clusters)) {
    cluster_number <- clusters[c]

    scresults[[t]][[c]]$results <- 
      FindMarkers(tumor, ident.1 = cluster_number) %>%
      arrange(-abs(avg_log2FC))
    labs <- head(scresults[[t]][[c]]$results, n = 15)

    plots[[c]][[t]]$volcano <-
      ggplot(scresults[[t]][[c]]$results, aes(avg_log2FC, -log10(p_val_adj))) +
        geom_point() +
        geom_label_repel(data = labs,
          aes(label = rownames(labs)),
          max.overlaps = 20) +
        labs(
          title = paste(c, "Genes"),
          subtitle = paste(t, "method"))
  }
}
```

# Perform GSEA using GO terms for BP, MP, and CC using DEGs

Identify pathways associated with changes in groups of those genes and plot results.

```{r}
degs <- list()

for(c in names(clusters)) {
  ds2[[c]]$results <-
    arrange(ds2[[c]]$results, desc(log2FoldChange))

  degs[["pb_ds2"]][[c]] <- as.vector(ds2[[c]]$results$log2FoldChange)
  names(degs[["pb_ds2"]][[c]]) <- rownames(ds2[[c]]$results)

  for(t in deg_test) {
    scresults[[t]][[c]]$results <-
      arrange(scresults[[t]][[c]]$results, desc(avg_log2FC))

    degs[[t]][[c]] <- as.vector(scresults[[t]][[c]]$results$avg_log2FC)
    names(degs[[t]][[c]]) <- rownames(scresults[[t]][[c]]$results)
  }
}

de_methods <- c("pb_ds2", "wilcox", "roc")
ont <- tribble(
  ~code, ~name,
  "BP", "GO BioProcesses",
  "MF", "GO MolFunction",
  "CC", "GO CellComponent"
)
ego <- list()

for(m in de_methods) {
  for(c in names(clusters)) {
    for(o in ont$code) {
    ego[[m]][[c]] <- clusterProfiler::gseGO(
        geneList = degs[[m]][[c]],
        OrgDb = org.Hs.eg.db::org.Hs.eg.db,
        ont = o,
        keyType = "SYMBOL",
        nPermSimple = 10000,
        eps = 0
      )

      ego[[m]][[c]] <- mutate(ego[[m]][[c]], p.adjust = -log10(p.adjust))

      plots[[c]][[m]][[o]] <- enrichplot::dotplot(ego[[m]][[c]],
        x = "NES",
        showCategory = 15) +
        labs(
          title = paste(c, "-", ont[ont$code == o, "name"]),
          subtitle = paste(m, "method")
        )
    }
  }
}
```

# Generate plots for presenation/publication

```{r}
# Export panels of DEG and GSEA analyses for display
for(c in names(clusters)) {
  for(m in de_methods) {
    plots[[c]][[m]]$panel <-
      (rDimPlot(tumor, "Osteosarcoma Primary Tumor Cells",
        subtitle = "Seurat-assigned clusters (optimized)") +
      plots[[c]][[m]]$volcano) /
      (plots[[c]][[m]]$BP +
      plots[[c]][[m]]$MF +
      plots[[c]][[m]]$CC)

    ggsave(paste0("HuOsteo/Plots/panel-", m, "-", c, ".pdf"),
      plots[[c]][[m]]$panel,
      device = "pdf",
      width = 18,
      height = 12,
      units = "in",)
  }
}

# Rename tumor clusters based on the above results
tumor <- RenameIdents(tumor,
  `0` = "Growth - recovery/ribogenesis",
  `1` = "Growth - actively cycling",
  `2` = "Growth - massive transcription",
  `3` = "Growth - massive translation",
  `4` = "Basal/quiescent",
  `5` = "Matrix producing/partially differentiated",
  `6` = "Inflammatory/interactive"
)

tumor$assignment <- Idents(tumor)
plots$assignment <-
  rDimPlot(tumor, "Osteosarcoma Primary Tumor Cells",
    subtitle = "Clusters assigned in this analysis",
    repel = T)

qs::qsave(tumor, "osteoshark.qs")

ggsave("HuOsteo/Plots/UMAP-new-labels.pdf",
  plots$assignment,
  device = "pdf",
  width = 6,
  height = 5,
  units = "in")

# Create a transcript density plot
plots$density <- rFeaturePlot(tumor, "nCount_RNA",
  title = "Transcript Density",
  subtitle = "Number of counts per cell",
  max.cutoff = 50000)

ggsave("HuOsteo/Plots/UMAP-count-density.pdf",
  plots$density,
  device = "pdf",
  width = 6,
  height = 5,
  units = "in")

# Create a plot of cell cyle predictions
tumor <- kill_cc(tumor)
plots$cell_cycle <- rDimPlot(tumor,
  title = "Cell Cycle Predictions",
  subtitle = "Based on transcriptional patterns")

ggsave("HuOsteo/Plots/UMAP-cell-cycle.pdf",
  plots$cell_cycle,
  device = "pdf",
  width = 6,
  height = 5,
  units = "in")
```
