## Cancer cells vs normal cells using SCEVAN
Single CEll Variational Aneuploidy aNalysis
When run, the result is in three metadata column:
class: with tumor (blue) vs normal celltype(green), and (filtered (red): low quality cells i think)
subclone: clones numbered like a seurat clusters
SCEVAN starts from the raw count matrix removing irrelevant genes and cells.
Identification of a small set of highly confident normal cells.
Relative gene expression obtained from removal of the baseline inferred from confident normal cells.
Edge-preserving nonlinear diffusion filtering of relative gene expression.
Segmentation with a variational region-growing algorithm.
Identification of normal cells as those in the cluster containing the majority of confident normal cells.
Segments are then classified in five copy number states.
Analysis of subclones including clone tree, pathway activities, but it can be turned off

```{r scevan}
#| fig.width: 10
#| fig.height: 10
#| fig.align: center
#| echo: FALSE
#| dependson: annot_celltypes
min_cells <- 500

for (group in c("patient_prim", "patient_mets", "mm_prim", "mm_mets")) {
    object <- qs::qread(str_c("output/seurat_objects/harmony_sobjs_annot/",
                              group,
                              ".qs"))

    count_mtx_list <- object %>%
        SplitObject(split.by = "sample_name") %>%
        lapply(FUN = function(x) GetAssayData(x, layer = "counts"))
    result_list <-
        parallel::mclapply(names(count_mtx_list),
                           mc.cores = min(1, parallelly::availableCores()),
                           function(x) {
            # If we have too few cells SCEVAN errors out
            # dim(x) must have a positive length
            if (ncol(count_mtx_list[[x]]) > min_cells) {
                output <-
                    SCEVAN::pipelineCNA(count_mtx = count_mtx_list[[x]],
                                        sample = x,
                                        organism = object$organism[1],
                                        par_cores = parallelly::availableCores(),
                                        SUBCLONES = FALSE)
                # SCEVAN writes a bunch of files to ./output/ so we need to move them
                system(paste0("mv output/",
                            x,
                            "* output/id_tumor/SCEVAN/"))
                # Two of the files have spaces in their names :-|
                system(paste0("mv output/' ",
                              x,
                              " vega_output' output/id_tumor/SCEVAN/"))
                system(paste0("mv output/' ",
                            x,
                            "onlytumor vega_output' output/id_tumor/SCEVAN/"))
                return(output)
            }
        })

    bind_rows(result_list) %>%
        mutate(scevan_tumor_call = if_else(class == "tumor",
                                           "cancer_cells",
                                           "normal_cells")) %>%
        rownames_to_column("cell") %>%
        write_tsv(str_c("output/id_tumor/SCEVAN/",
                        group,
                        "_metadata.tsv"))
}
```

## Cancer cells vs normal cells using scATOMIC
The result comes with the annotation of cell types including for both cancer type
and the normal cell types. Also outputs for each cell:
Classification_confidence: confident, low_confidence on annotation
pan_cancer_cluster: either cancer or normal
scATOMIC_pred : celltype annotation
```{r scatomic}
#| fig.width: 10
#| fig.height: 10
#| fig.align: center
#| echo: FALSE
#| dependson: annot_celltypes
library(scATOMIC)

for (group in c("patient_prim", "patient_mets")) {
    object <- qs::qread(str_c("output/seurat_objects/harmony_sobjs_annot/",
                              group,
                              ".qs"))

    min_count_per_gene <- 10
    count_mtx_list <- object %>%
        SplitObject(split.by = "sample_name") %>%
        lapply(FUN = function(x) {
            temp_data <- GetAssayData(x, layer = "counts")
            temp_data <- temp_data[rowSums(temp_data) > min_count_per_gene, ]
        })

    result_list <-
        parallel::mclapply(names(count_mtx_list),
                           mc.cores = parallelly::availableCores(),
                           function(x) {
            cell_predictions <-
                scATOMIC::run_scATOMIC(
                    rna_counts = count_mtx_list[[x]],
                    mc.cores = parallelly::availableCores()
                )
            scATOMIC_results <-
                scATOMIC::create_summary_matrix(
                    prediction_list = cell_predictions,
                    use_CNVs = FALSE,
                    modify_results = TRUE,
                    mc.cores = parallelly::availableCores(),
                    raw_counts = count_mtx_list[[x]],
                    min_prop = 0.5
                ) %>%
                select(starts_with("layer_"),
                       "scATOMIC_pred",
                       "classification_confidence")
        })

    bind_rows(result_list) %>%
    mutate(scatomic_tumor_call = if_else(grepl("Cancer Cell$", scATOMIC_pred),
                                         "cancer_cells",
                                         "normal_cells")) %>%
    rownames_to_column("cell") %>%
    write_tsv(str_c("output/id_tumor/scATOMIC/",
                    group,
                    "_metadata.tsv"))
}
```
